diff -up firefox-53.0/browser/app/profile/firefox.js.addons firefox-53.0/browser/app/profile/firefox.js
--- firefox-53.0/browser/app/profile/firefox.js.addons	2017-04-18 10:03:02.300234728 +0200
+++ firefox-53.0/browser/app/profile/firefox.js	2017-04-18 10:05:38.170676613 +0200
@@ -63,7 +63,8 @@ pref("extensions.systemAddon.update.url"
 
 // Disable add-ons that are not installed by the user in all scopes by default.
 // See the SCOPE constants in AddonManager.jsm for values to use here.
-pref("extensions.autoDisableScopes", 15);
+pref("extensions.autoDisableScopes", 0);
+pref("extensions.showMismatchUI", false);
 
 // Whether or not webextension themes are supported.
 pref("extensions.webextensions.themes.enabled", false);
