diff -up firefox-55.0/js/src/jit/AtomicOperations.h.1253216 firefox-55.0/js/src/jit/AtomicOperations.h
--- firefox-55.0/js/src/jit/AtomicOperations.h.1253216	2017-08-02 11:00:24.115991657 +0200
+++ firefox-55.0/js/src/jit/AtomicOperations.h	2017-08-02 11:01:17.373817096 +0200
@@ -348,7 +348,7 @@ AtomicOperations::isLockfree(int32_t siz
 # elif defined(__sh__)
 #  include "jit/none/AtomicOperations-feeling-lucky.h"
 # else
-#  include "jit/none/AtomicOperations-none.h" // These MOZ_CRASH() always
+#  include "jit/none/AtomicOperations-ppc.h"
 # endif
 #elif defined(JS_CODEGEN_X86) || defined(JS_CODEGEN_X64)
 # include "jit/x86-shared/AtomicOperations-x86-shared.h"
